File: day09/puzzle.js
import { createAocPuzzle } from "./aoc.js";

const testAnswers = {
  part1: 15,
  part2: 1134
};

function parseInput(lines) {
  return lines.map(line => line.split("").map(point => +point));
}

function runPart1(input) {
  const maxX = input[0].length - 1;
  const maxY = input.length - 1;
  let riskLevel = 0;
  for (let y = 0; y <= maxY; ++y) {
    for (let x = 0; x <= maxX; ++x) {
      const point = input[y][x];
      if (
        (x == 0 || point < input[y][x - 1]) &&
        (y == 0 || point < input[y - 1][x]) &&
        (x == maxX || point < input[y][x + 1]) &&
        (y == maxY || point < input[y + 1][x])
      ) {
        riskLevel += point + 1;
      }
    }
  }
  return riskLevel;
}

function runPart2(input) {
  const width = input[0].length;
  const height = input.length;
  const maxX = width - 1;
  const maxY = height - 1;
  const checkedPositions = new Int8Array(width * height);

  function getBasinSize(x, y) {
    if (checkedPositions[y * width + x] || input[y][x] == 9) {
      return 0;
    }

    checkedPositions[y * width + x] = 1;
    let size = 1;
    if (x > 0) {
      size += getBasinSize(x - 1, y);
    }
    if (y > 0) {
      size += getBasinSize(x, y - 1);
    }
    if (x < maxX) {
      size += getBasinSize(x + 1, y);
    }
    if (y < maxY) {
      size += getBasinSize(x, y + 1);
    }

    return size;
  }

  const lowPoints = [];
  for (let y = 0; y <= maxY; ++y) {
    for (let x = 0; x <= maxX; ++x) {
      const point = input[y][x];
      if (
        (x == 0 || point < input[y][x - 1]) &&
        (y == 0 || point < input[y - 1][x]) &&
        (x == maxX || point < input[y][x + 1]) &&
        (y == maxY || point < input[y + 1][x])
      ) {
        lowPoints.push({ x, y });
      }
    }
  }

  const basinSizes = new Array(lowPoints.length);
  for (let i = 0; i < lowPoints.length; ++i) {
    basinSizes[i] = getBasinSize(lowPoints[i].x, lowPoints[i].y);
  }
  basinSizes.sort((a, b) => b - a);

  console.log("Three largest basins:", basinSizes[0], basinSizes[1], basinSizes[2]);
  return basinSizes[0] * basinSizes[1] * basinSizes[2];
}

export function createPuzzle() {
  return createAocPuzzle({ parseInput, runPart1, runPart2, testAnswers });
}

File: day13/puzzle.py
from aoc import AocPuzzle


TEST_ANSWERS = { "part1": 17 }


class FoldInstruction:
    def __init__(self, instruction: str):
        parts = instruction.split("=")
        self.fold_along = parts[0][-1]
        self.line = int(parts[1])


class PageOne:
    def __init__(self, dots: list[str], fold_instructions: list[str]):
        coords: list[list[int]] = []
        max_x = 0
        max_y = 0
        for dot in dots:
            coord = [int(coord) for coord in dot.split(",")]
            coords.append(coord)
            if coord[0] > max_x:
                max_x = coord[0]

            if coord[1] > max_y:
                max_y = coord[1]

        self.paper = [[False for _ in range(max_x + 1)] for _ in range(max_y + 1)]
        for coord in coords:
            self.paper[coord[1]][coord[0]] = True

        self.fold_instructions = [FoldInstruction(instruction) for instruction in fold_instructions]


def fold_paper(paper: list[list[bool]], instruction: FoldInstruction) -> list[list[bool]]:
    match instruction.fold_along:
        case "x":
            return [[row[x] | row[-x - 1] for x in range(instruction.line)] for row in paper]
        case "y":
            return [[paper[y][x] | paper[-y - 1][x] for x in range(len(paper[0]))] for y in range(instruction.line)]


def print_paper(paper: list[list[bool]]):
    for row in paper:
        print("".join(["#" if dot else "." for dot in row]))


class Puzzle(AocPuzzle[PageOne, int]):
    def __init__(self):
        super().__init__(TEST_ANSWERS)


    def parse_input(self, lines: list[str]):
        dots: list[str] = []
        i = 0
        while lines[i].rstrip():
            dots.append(lines[i])
            i += 1

        return PageOne(dots, lines[i + 1:])


    def run_part1(self):
        paper = fold_paper(self._input.paper, self._input.fold_instructions[0])
        if self._is_test:
            print_paper(paper)

        return sum([sum(row) for row in paper])


    def run_part2(self):
        paper = self._input.paper
        for fold_instruction in self._input.fold_instructions:
            paper = fold_paper(paper, fold_instruction)

        print_paper(paper)
